packages/vite/src/node/plugins/esbuild.ts

```diff
@@ -243,20 +258,55 @@ export const buildEsbuildPlugin = (config: ResolvedConfig): Plugin => {
         return null
       }
 
-      const res = await transformWithEsbuild(code, chunk.fileName, {
+      const isEsLibBuild = config.build.lib && opts.format === 'es'
+      let options: TransformOptions = {
         ...config.esbuild,
         target: target || undefined,
-        ...(minify
-          ? {
-              // Do not minify ES lib output since that would remove pure annotations
-              // and break tree-shaking
-              // https://github.com/vuejs/core/issues/2860#issuecomment-926882793
-              minify: !(config.build.lib && opts.format === 'es'),
-              treeShaking: true,
-              format: rollupToEsbuildFormatMap[opts.format]
-            }
-          : undefined)
-      })
+        format: rollupToEsbuildFormatMap[opts.format]
+      }
+
+      if (minify) {
+        // If user enable fine-grain minify options, minify with their options instead
+        if (
+          options.minifyIdentifiers === true ||
+          options.minifySyntax === true ||
+          options.minifyWhitespace === true
```

Comment:
```suggestion
          options.minifyIdentifiers ||
          options.minifySyntax ||
          options.minifyWhitespace
```
Maybe we could simplify this a bit?